File: testapp/src/main/java/com/example/MyJavalinServlet.java
package com.example;

import io.javalin.Javalin;
import io.javalin.http.JavalinServlet;

import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

@WebServlet(name = "MyJavalinServlet", urlPatterns = {"/rest/*"})
public class MyJavalinServlet extends HttpServlet {
    private final JavalinServlet javalin = Javalin.createStandalone()
            .get("/rest", ctx -> ctx.result("Hello!"))
            .javalinServlet();

    @Override
    protected void service(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
        javalin.service(req, resp);
    }
}

File: testapp-kotlin-tomcat/src/main/kotlin/com/example/MyJavalinServlet.kt
package com.example

import io.javalin.Javalin
import io.javalin.http.Context
import io.javalin.http.servlet.JavalinServlet
import jakarta.servlet.annotation.WebServlet
import jakarta.servlet.http.HttpServlet
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse

@WebServlet(name = "MyJavalinServlet", urlPatterns = ["/rest/*"])
class MyJavalinServlet : HttpServlet() {
    private val javalin: JavalinServlet = Javalin.createStandalone()
        .get("/rest") { ctx: Context -> ctx.result("Hello!") }
        .javalinServlet()

    override fun service(req: HttpServletRequest, resp: HttpServletResponse) {
        javalin.service(req, resp)
    }
}
